Michael Uftring   
Indiana University   
Network Science, Spring 2018   
Week#8 Assignment - Link Communities   

# Link communities
For this assignment we will take a look at link communities and how they differ from node communities. In order to do this we will use the algorithm discussed in the reading ("Link communities reveal multiscale complexity in networks") and link community video from canvas.

A small python module has been prepared that will allow you to use the link community algorithm with Anaconda and Python 3.5. To install the module you want to open a terminal or shell and use:

```
pip install git+https://github.com/Nathaniel-Rodriguez/linkcom.git
```

This will install the package locally in your Anaconda site-packages directory (the same place where the `conda` command would install new packages). Make sure you have [git](https://git-scm.com/) installed first. If you are using Windows you will need to use the Anaconda command prompt when using pip, so that it adds the package to Anaconda. If you have trouble installing the package you can just unpack the zip file from the [github repository](https://github.com/Nathaniel-Rodriguez/linkcom/tree/master/linkcom) and put the linkcom folder in your working directory.

To use the package you can do:

In [1]:
import linkcom

The code has been adapted so that it will work with networkx graphs. The link communities algorithm requires simple undirected graphs to use. This means there can't be any self-loops or parallel edges. However, you can use weighted graphs.

In [2]:
# First lets import networkx
import networkx as nx

# And generate a new graph
my_graph = nx.erdos_renyi_graph(100, 0.1)

# We need to make sure this is a graph of type Graph
print(type(my_graph).__name__)

Graph


If the graph you load in isn't of type Graph (it maybe a Multigraph or DiGraph), it is easy to convert it to one:

In [3]:
my_graph = nx.Graph(my_graph)

## Using linkcom
Now lets call the `cluster` method in `linkcom` to cluster the links of the graph. The `cluster` method takes several optional arguments:

```
linkcom.cluster(nx_graph, threshold=None, is_weighted=False, weight_key='weight', dendro_flag=False, to_file=False, basename="clustering", delimiter='\t')
```

`Threshold`: sets the cut-off for the dendrogram.

`is_weighted`: can be `True` or `False` depending upon whether the graph has weights or not. Set this to `True` if the graph is weighted.

`weight_key`: specifies what attribute the edges have that has weight values. In networkx it is convention that this key be set to `weight`. Most functions in networkx will assume this is the key. This is also the default value for the cluster method.

`dendro_flag`: specifies whether to return the dendrogram (only applicable if the graph is unweighted and no threshold is given).

`to_file`: specifies whether to write the outputs to file. Several files will be written and given names based on `basename` with elements separated by `delimiter`. 

These outputs will be written to file:

```
Three text files with extensions .edge2comm.txt, .comm2edges.txt,
and .comm2nodes.txt store the communities.

edge2comm, an edge on each line followed by the community
id (cid) of the edge's link comm:
node_i <delimiter> node_j <delimiter> cid <newline>

comm2edges, a list of edges representing one community per line:
cid <delimiter> ni,nj <delimiter> nx,ny [...] <newline>

comm2nodes, a list of nodes representing one community per line:
cid <delimiter> ni <delimiter> nj [...] <newline>

The output filename contains the threshold at which the dendrogram
was cut, if applicable, or the threshold where the maximum
partition density was found, and the value of the partition 
density.

If no threshold was given to cut the dendrogram, a file ending with
'_thr_D.txt' is generated, containing the partition density as a
function of clustering threshold.

If the dendrogram option was given, two files are generated. One with
'.cid2edge.txt' records the id of each edge and the other one with
'.linkage.txt' stores the linkage structure of the hierarchical 
clustering. In the linkage file, the edge in the first column is 
merged with the one in the second at the similarity value in the 
third column.
```

The cluster method will return a tuple with different elements:

    If no threshold is given, then a tuple is returned with: (dict) dictionary with keys=edges and values=community membership, (float) best similarity, (float) best partition density, (list) partition density list.

    If dendro_flag is given (only applicable if no threshold), then a tuple is returned with: (dict) dictionary with keys=edges and values=community membership, (float) best similarity, (float) best partition density, (list) partition density list, (dict) keys=edges and values=community membership for original, (list) dendrogram.

    If threshold is given a tuple is returned with: (dict) dictionary with keys=edges and values=community membership, partition density at threshold.

You will mostly just be interested in using the dictionary which contains the community assignment data, which is always the first element of the tuple. Lets do a short example:

In [4]:
e2c, S, D, Dlist = linkcom.cluster(my_graph)

clustering...
computing similarities...
# D_max = 0.082045
# S_max = 0.192308


If we print `e2c` we will see that each edge has a community membership:

In [5]:
print(e2c)

{(26, 47): 106, (32, 54): 350, (82, 86): 249, (12, 78): 39, (89, 96): 284, (22, 44): 266, (63, 84): 79, (36, 87): 451, (34, 35): 248, (73, 82): 3, (9, 51): 350, (41, 57): 4, (43, 64): 451, (23, 90): 436, (55, 80): 7, (6, 72): 106, (28, 75): 255, (77, 89): 106, (6, 95): 296, (56, 88): 439, (19, 76): 241, (73, 81): 297, (68, 92): 8, (10, 75): 9, (25, 49): 211, (24, 39): 261, (5, 11): 450, (7, 9): 350, (38, 96): 263, (20, 32): 39, (3, 40): 13, (1, 40): 13, (23, 53): 451, (22, 39): 266, (62, 94): 106, (27, 40): 268, (8, 38): 16, (7, 22): 350, (6, 26): 106, (55, 56): 106, (61, 69): 19, (10, 41): 270, (8, 57): 271, (16, 38): 21, (39, 85): 24, (4, 73): 106, (60, 62): 106, (50, 79): 149, (40, 54): 350, (3, 74): 123, (61, 83): 451, (40, 73): 27, (37, 57): 97, (28, 77): 106, (4, 98): 106, (56, 95): 106, (69, 73): 29, (16, 75): 30, (17, 83): 31, (66, 72): 106, (56, 69): 280, (18, 81): 33, (0, 49): 325, (92, 98): 106, (46, 65): 106, (70, 82): 172, (66, 98): 106, (21, 49): 286, (4, 31): 287, (3, 17

Since this is a random graph, we expect there not to be any meaningful communities, and indeed they are all labelled `56` (the community ID), so there doesn't appear to be any link communities in the graph.

We can now readily take our results and put them back into our graph so that it can be saved and viewed in Gephi. We can do this using the [`set_edge_attributes`](https://networkx.github.io/documentation/stable/reference/generated/networkx.classes.function.set_edge_attributes.html?highlight=set_edge_attributes#networkx.classes.function.set_edge_attributes) function in networkx. It works just like the `set_node_attributes` function from previuos assignments, but with edges instead:

In [6]:
# Put the link communities into the graph
# Note this function's syntax depends on the networkx version.
# If you use networkx 1.9 and below this line should be
# nx.set_edge_attributes(my_graph, "linkcom", e2c)
nx.set_edge_attributes(my_graph, e2c, "linkcom")

# Save the graph to file
nx.write_gexf(my_graph, "my_graph.gexf")

TypeError: unhashable type: 'dict'

In [10]:
nx.set_edge_attributes(my_graph, "linkcom", e2c)

In [11]:
my_graph[0]

{1: {'linkcom': 449},
 3: {'linkcom': 449},
 15: {'linkcom': 325},
 42: {'linkcom': 151},
 43: {'linkcom': 421},
 48: {'linkcom': 467},
 49: {'linkcom': 325},
 56: {'linkcom': 106},
 62: {'linkcom': 106},
 66: {'linkcom': 106},
 72: {'linkcom': 106},
 77: {'linkcom': 106},
 88: {'linkcom': 439}}

In [12]:
# Save the graph to file
nx.write_gexf(my_graph, "my_graph.gexf")

Now if we were to open the graph in Gephi we should be able to use the link communities to color the graph edges. In Gephi you may need to click the little attribute type button so that edge attributes are set to [ranked rather than numeric](https://gephi.org/tutorials/gephi-tutorial-quick_start.pdf). This is because we want to color the edges according to their membership and not with a gradient. Additionally, in order to keep the edge colors when saving the graph you will need to make sure the edge color in the Preview tab is set to `original`. Lastly, since you will be looking at link communities (which determine the node membership in link clustering) it will be helpful to increase the size of the edges in Gephi so the colors are more visible and so you can detect nodes that belong to multiple communities.

## The network science collaboration graph
You will be using the NetSci collaboration graph for your assignment. The nodes of the graph are people and links are formed between people who co-author a scientific paper together in network science. You can download it from [here](http://vlado.fmf.uni-lj.si/pub/networks/data/collab/netscience.htm). Make sure to read the graph description on that page. The graph has ~1500 nodes and is partly disconnected. If you have difficulty working with the full graph or trouble loading it into Gephi, you can use the largest connected component (which only has about 350 nodes). The largest connected component of a graph can be returned from networkx using:

```
largest_component=max(nx.connected_component_subgraphs(my_graph), key=len)
```

Most of the interesting stuff is happening around this component anyway. So you don't lose much except for the scientists and groups that decided to work alone.

Follow these steps for the assignment:

1. Load the Netsci graph and run the link communities clustering algorithm on it.
2. Save the link communities to the graph and save the graph to file.
3. In Gephi choose a good layout for the graph.
4. Run the modularity command to generate communities for the nodes.
5. Color the edges according to the link communities and the nodes according to the communities found by Gephi. Remember to take care in choosing the resolution parameter.
6. How well does link clustering do at detecting community structure? How well does Gephi's node modularity do at detecting community structure? What do you think the communities represent?
7. What are the similarities and differences between the communities detected by either algorithm?
8. Which authors have a prominent position in multiple communities? What do you think these author's roles are?
9. What other features do you notice about the graph that are captured with overlaping communities?
10. Save your visualization to file.
11. Once complete, submit a PDF document to Canvas that contains your responses and your graph visualization (since this is a larger graph feel free to crop the figure so that it only includes parts relevant to your responses).

## Answers

#### 1. Load the Netsci graph and run the link communities clustering algorithm on it

In [28]:
graph = nx.read_pajek("netsience.net")

In [29]:
print(nx.info(graph))

Name: 
Type: MultiGraph
Number of nodes: 1589
Number of edges: 2742
Average degree:   3.4512


In [30]:
graph = nx.Graph(graph)

In [31]:
print(nx.info(graph))

Name: 
Type: Graph
Number of nodes: 1589
Number of edges: 2742
Average degree:   3.4512


In [32]:
e2c, S, D, Dlist = linkcom.cluster(graph)

clustering...
computing similarities...
# D_max = 0.693791
# S_max = 0.368421


#### 2. Save the link communities to the graph and save the graph to file.

In [33]:
nx.set_edge_attributes(graph, "linkcom", e2c)

In [34]:
nx.write_gexf(graph, "network-science-collaboration-network.gexf")

#### 3. In Gephi choose a good layout for the graph.

#### 4. Run the modularity command to generate communities for the nodes.

[Gephi - Clustering layout by modularity](http://parklize.blogspot.com/2014/12/gephi-clustering-layout-by-modularity.html)

#### 5. Color the edges according to the link communities and the nodes according to the communities found by Gephi. Remember to take care in choosing the resolution parameter.

#### 6. How well does link clustering do at detecting community structure? How well does Gephi's node modularity do at detecting community structure? What do you think the communities represent?

comments

#### 7. What are the similarities and differences between the communities detected by either algorithm?

comments

#### 8. Which authors have a prominent position in multiple communities? What do you think these author's roles are?

comments

#### 9. What other features do you notice about the graph that are captured with overlaping communities?

comments

#### 10. Save your visualization to file.

#### 11. Once complete, submit a PDF document to Canvas that contains your responses and your graph visualization (since this is a larger graph feel free to crop the figure so that it only includes parts relevant to your responses).